Store Orders: List coupons in the totals section

ryelle opened the pull request:
This PR adds the applied coupons to the single order view, in place of the "Discounts" total label, to imply "these coupons discounted this amount". Also it fits better this way 😁 

<img width="608" alt="screen shot 2017-12-18 at 5 47 22 pm" src="https://user-images.githubusercontent.com/541093/34132029-0f512ff4-e41c-11e7-9c0e-bb4188bc78f6.png">

I've also added parenthesis around the discount total, because technically this amount has already been subtracted from the order subtotal (each product shows the post-coupon price). The discount total here is more of a reference number than a subtotal to be tallied. 

This does not touch the edit-state yet. I'm marking this as **In Progress** so that I can keep working there, but I'd like some designer input on whether this positioning is OK, @kellychoffman @jameskoster 

**To test**

1. Create an order with at least one coupon applied
2. View the order in calypso: `/store/order/:site/:orderId`
3. You should see the coupon applied, and the total discount granted
4. If we can tell which coupon this is, we'll link the code to the promotions edit screen - coupons applied via editing in wp-admin don't include the coupon ID in meta-data, so we can't create the link for those. given most orders are not created via wp-admin, this seems okay.

kellychoffman commented:
This looks nice! 

The only request I have is to keep the left padding the same as the above products and also to add a bottom-border to separate it from the totals section.

jameskoster commented:
> The only request I have is to keep the left padding the same as the above products and also to add a bottom-border to separate it from the totals section.

Done. 

@ryelle could we hide the coupons row when no coupon has been applied?

ryelle commented:
>  could we hide the coupons row when no coupon has been applied?

Sure, I've updated the logic for that. I also cleaned up the CSS a little, since it wasn't correctly applying the border on orders with tax values. So it looks like this now:

<img width="594" alt="screen shot 2017-12-20 at 12 34 10 pm" src="https://user-images.githubusercontent.com/541093/34220345-519bf22e-e582-11e7-8deb-97f17e7776ee.png">

ryelle commented:
This latest update adds UI for removing coupons from an existing order, and looks like this:

<img width="598" alt="edit-coupon" src="https://user-images.githubusercontent.com/541093/34221443-08e61984-e586-11e7-875b-5accf38b9080.png">

Unfortunately, due to an API issue https://github.com/woocommerce/woocommerce/pull/18236, removing a coupon doesn't actually work correctly (we can still merge this, probably, but it won't update the totals correctly and can be confusing to test). Going to switch this to Needs Review just to get the code reviewed, and hopefully we can get the API side worked out soon.

ryelle commented on client/extensions/woocommerce/app/order/order-details/table.js:
```diff
@@ -261,6 +279,85 @@ class OrderDetailsTable extends Component {
+		if ( isEditing ) {
+			const couponMarkup = coupons.map( item => {
+				if ( ! item.code ) {
+					return null;
```
Nope, in fact a null code is how you delete a coupon.

ryelle commented on client/extensions/woocommerce/app/order/order-details/table.js:
```diff
@@ -261,6 +279,85 @@ class OrderDetailsTable extends Component {
+		}
+
+		if ( isEditing ) {
+			const couponMarkup = coupons.map( item => {
```
So... technically no, since this section returns at L306, these two `couponMarkup`s never overlap, but it does make for a double-take while reading the code, so I'll make the change.

allendav commented:
Like @timmyc I too took pause at using parens - I'm used to seeing it in accounting to reflect a negative amount, and I kinda want that column of numbers you're presenting to add up to the total.

If we want to show the coupon amounts like this, how about we show undiscounted item amounts so these things add up correctly?

allendav commented on client/extensions/woocommerce/app/order/order-details/table.js:
```diff
@@ -261,6 +279,85 @@ class OrderDetailsTable extends Component {
+
+			return (
+				<TableRow className="order-details__coupon-list">
+					<TableItem className="order-details__coupon-list-label" colSpan={ showTax ? 2 : 1 }>
```
Nice trick with the colSpan

allendav commented on client/extensions/woocommerce/app/order/order-details/table.js:
```diff
@@ -261,6 +279,85 @@ class OrderDetailsTable extends Component {
+						<div className="order-details__coupon-list-tokens">{ couponMarkup }</div>
+					</TableItem>
+					<TableItem className="order-details__totals-value">
+						({ formatCurrency( order.discount_total, order.currency ) })
```
Why the outer parens?

allendav replied:
LOL. Nevermind. :P

jameskoster commented:
Tweaked a couple of style issues I noticed. I tend to agree on the parenthesis as well fwiw.

ryelle commented:
I didn't know that about parens having a meaning in accounting, but now I do 🙂  How do you suggest we indicate that the value isn't part of the totals, then? As it is now, if you try to go down the totals col and add in the negative discount, you'll be applying the discount twice (because the product totals already account for it).

timmyc commented:
I suppose when in doubt, let's look to core:

![coupon](https://user-images.githubusercontent.com/22080/34532677-dd9a0718-f06b-11e7-8297-939f82c2b98c.png)

No parens there, and yes things don't fully ad up in the column either - but my accounting 101 self can get past that 😆 

Unless @jameskoster has other thoughts, I say we drop the parens and get this merged in

ryelle commented:
I've rebased this & removed the parens, should be good for a final 👀

kellychoffman commented:
>  I say we drop the parens and get this merged in

Agree. :shipit:

ryelle commented:
Forgot to add the screenshot: ![screen shot 2018-01-03 at 1 48 02 pm](https://user-images.githubusercontent.com/541093/34534611-ca3fc5d8-f08c-11e7-8de6-8aee8d6f25de.png)

kellychoffman commented:
Looking again, it feels like an additional charge. Any objections to including a `-` in front of the coupon amounts?

ryelle commented:
Negative sign added:

![screen shot 2018-01-03 at 2 27 43 pm](https://user-images.githubusercontent.com/541093/34536170-58e0d534-f092-11e7-938a-3118c8beb9a6.png)

kellychoffman commented:
I think this feels better. I know there's the overall math issue, but it still feels like a discount indicator instead of a charge. I say :shipit:

jameskoster commented:
Tax column looks out of alignment. On that note, numerical values should probably be right-justified. But we don't have to address that here.

ryelle commented:
@timmyc @kellychoffman Since removing coupons is still [a little ways off](https://github.com/woocommerce/wc-api-dev/issues/78), should I split that out of this PR, or can it just merge as-is? (editing an order is getting pretty low traffic, but removing a coupon doesn't currently do anything to the final price)

timmyc commented:
I think we are fine with letting this merge in as-is, and can circle back and change up items if it becomes a problem.

kellychoffman commented:
I don't think that removing coupons is a must have for this first version of order editing. Let's circle back after its been live a while and we've done some testing.

ryelle commented:
@kellychoffman Removing and adding are basically going to happen at the same time, so technically it's not really a problem (unless you're saying specifically we don't want removal?). the issue is that this PR already does the removing UI, but removing a coupon doesn't actually do anything to the total.

kellychoffman commented:
Oh I see. I misunderstood. Carry on :)

timmyc commented:
@ryelle and I chatted a bit more about this, and [her original suggestion](https://github.com/Automattic/wp-calypso/pull/20929#issuecomment-355322092) is definitely looking like the right way to go now since we are blocked on the coupon removal front.

ryelle commented:
I split out the removal into a new branch, #21280. This could use one more ✅  since I did change code.

I noticed after rebasing that the totals section is misaligned now - the divider doesn't go fully across:

![screen shot 2018-01-04 at 3 53 03 pm](https://user-images.githubusercontent.com/541093/34583907-93a9da82-f167-11e7-8ef3-731cbb4d409b.png)

@jameskoster I'm not sure how you want to fix this, since I think the extra padding was introduced in your style cleanup. I'm leaning to merging this and fixing the alignment later, since this one's gone back and forth a bunch.

timmyc commented on client/extensions/woocommerce/app/order/order-details/table.js:
```diff
@@ -133,19 +133,36 @@ class OrderDetailsTable extends Component {
+		}
+	};
+
+	removeCoupon = item => () => {
```
So we are leaving this here for now even though it isn't being used until #21280 lands

ryelle replied:
No, I just missed it 🤦‍♀️ – removed in 45b06df80ce2895b07f123ecad958e71f14ff257

timmyc commented:
Thanks for fixing that up @jameskoster - looks great now

![allign](https://user-images.githubusercontent.com/22080/34613823-14008cd6-f1e4-11e7-880a-9d4b4f032fc3.png)
